[WIP] Merlin Inference API Example Notebooks (#645)
NVIDIA/NVTabular: examples/inference_triton/inference-HugeCTR/README.md

# Inference with HugeCTR Model

In this folder, we provide two example notebooks, [movielens-HugeCTR](https://github.com/NVIDIA/NVTabular/tree/main/examples/inference_triton/inference-HugeCTR/movielens-HugeCR.ipynb) and [movielens-HugeCTR-inference](https://github.com/NVIDIA/NVTabular/tree/main/examples/inference_triton/inference-HugeCTR/movielens-inference.ipynb), and explain the steps to do inference with Merlin Inference API after training a model with HugeCTR framework. 

## Getting Started 

> In order to use Merlin Inference API, there are two containers that the user needs to build and launch. The first one is for preprocessing with NVTabular and training a model with HugeCTR framework. The other one is for serving/inference. 

Reviewer: ```suggestion
There are two containers that are needed in order to use the Merlin Inference API. The first one is for preprocessing with NVTabular and training a model with HugeCTR framework. The other one is for serving/inference. 
```

## 1. Pulling the NVTabular Docker Container:

> We start with pulling NVTabular container. This is to do preprocessing, feature engineering on our datasets using NVTabular, and then to train a DL model with HugeCTR framework with processed datasets.

Reviewer: ```suggestion
We start with pulling the Merlin training container. This is to do preprocessing, feature engineering on our datasets using NVTabular, and then to train a DL model with HugeCTR framework with processed datasets.
```

Before starting docker continer, first create a `nvt_triton` directory and `data` subdirectory on your host machine:

```
mkdir -p nvt_triton/data/
cd nvt_triton
```
We will mount `nvt_triton` directory into the NVTabular docker container.

Merlin containers are available in the NVIDIA container repository at the following location: http://ngc.nvidia.com/catalog/containers/nvidia:nvtabular.

You can pull the `Merlin-Training` container by running the following command:

```

> docker run --gpus=all -it -v ${PWD}:/model/ -p 8888:8888 -p 8797:8787 -p 8796:8786 --ipc=host nvcr.io/nvstaging/merlin/merlin-training:0.4 /bin/bash

Reviewer: Lets go with the release version
```suggestion
docker run --gpus=all -it -v ${PWD}:/model/ -p 8888:8888 -p 8797:8787 -p 8796:8786 --ipc=host nvcr.io/nvidia/merlin/merlin-training:0.4 /bin/bash
```

```
The container will open a shell when the run command execution is completed. You'll have to start the jupyter lab on the Docker container. It should look similar to this:

```
root@2efa5b50b909:
```

Activate the rapids conda environment by running the following command:
```
root@2efa5b50b909: source activate rapids
```
You should receive the following response, indicating that the environment has been activated:

```
(rapids)root@2efa5b50b909:
```

> 1) Install Triton Python Client Library:

Reviewer: Isn't the triton python client library installed on the containers? Can we delete this section

- do preprocessing with NVTabular
- serialize and save a workflow to load later to transform new dataset
- train a HugeCTR MLP model and save it in the `/models` directory.

The following notebook [movielens-HugeCTR-inference](https://github.com/NVIDIA/NVTabular/tree/main/examples/inference_triton/movielens_inference.ipynb) shows how to send request to Triton IS 
- to transform new data with NVTabular
- to generate prediction results for new dataset.

Now you can start `movielens-HugeCTR` and `movielens-HugeCTR-inference` notebooks. Note that you need to save your workflow and DL model in the `models` directory before launching the `tritonserver` as defined below. Then you can run the `movielens-HugeCTR-inference` example notebook once the server is started.

## 3. Build and Run the Triton Inference Server container:

1) Navigate to the `nvt_triton` directory that you saved NVTabular workflow, Tensorflow model and the ensemble model.
```
cd <path to nvt_triton>
```

2) Launch Merlin Triton Inference Server container:
```

> docker run -it --gpus=all --shm-size=1g --ulimit memlock=-1 --ulimit stack=67108864 -p 8000:8000 -p 8001:8001 -p 8002:8002 -v ${PWD}:/model/ nvcr.io/nvstaging/merlin/merlin-inference:0.4

Reviewer: ```suggestion
docker run -it --gpus=all --shm-size=1g --ulimit memlock=-1 --ulimit stack=67108864 -p 8000:8000 -p 8001:8001 -p 8002:8002 -v ${PWD}:/model/ nvcr.io/nvidia/merlin/merlin-inference:0.4
```
